src/mawpy/steps/incremental_clustering.py

```diff
@@ -0,0 +1,396 @@
+        cluster_center = [
+            str(np.mean([each_point[0] for each_point in this_cluster.pList])),
+            str(np.mean([each_point[1] for each_point in this_cluster.pList])),
+        ]  # calculate center
+        for each_point in this_cluster.pList:  # for each coordinate in c.pList
+            locations_to_cluster_center_map[(str(each_point[0]), str(each_point[1]))] = (
+                cluster_center[0],
+                cluster_center[1],
+                cluster_radius,
+            )  # store the center with cluster radius
+    return locations_to_cluster_center_map
+
+
+def _run_for_user(df_by_user, spat_constr, dur_constr=None):
+
+    """
+        Function to perform incremental clustering on a dataframe containing traces for a single user
+    """
+
+    if dur_constr:  # cluster locations of stays to obtain aggregated stayes
+        # get unique GPS stay points if stay duration is greater than duration constraint
+
+        locations_for_clustering = ((df_by_user[df_by_user[STAY_DUR] >= dur_constr]
+                                     .apply(lambda x: (x[STAY_LAT], x[STAY_LONG]), axis=1))
+                                    .tolist())
+    else:  # cluster original locations (orig_lat and orgi_long) to obtain stays
+        # get GPS original points
+        locations_for_clustering = (df_by_user.apply(lambda x: (x[ORIG_LAT], x[ORIG_LONG]), axis=1)).tolist()
+
+    if len(locations_for_clustering) == 0:
+        return df_by_user
+
+    clusters_list = _get_clusters(locations_for_clustering, spat_constr)
+    # make Ccurrent as Cnew
+
+    ### apply k means clustering with k same as length of L
+    clusters_list = _k_means_cluster_lloyd(
+        clusters_list
+    )  # correct an order issue related to incremental clustering # clusters get appended to list L
+
+    ### create a dictionary which takes each point and keeps information of its cluster center and radius
+    ## centers of each locations that are clustered
+    locations_to_cluster_center_map = _get_locations_to_cluster_center_map(clusters_list)
+
+    ### Update trace itself using clustre center and max(radius, uncertainty)
+    df_by_user[[STAY_LAT, STAY_LONG, STAY_UNC]] = df_by_user.apply(
+        lambda row: _get_cluster_center(row, locations_to_cluster_center_map, dur_constr), axis=1, result_type='expand')
+
+    df_by_user = df_by_user.sort_values(by=[UNIX_START_DATE], ascending=True)
+
+    df_by_user[STAY] = ((df_by_user[[STAY_LAT, STAY_LONG]] != df_by_user[[STAY_LAT, STAY_LONG]].shift())
+                          .any(axis=1).cumsum())
+
+    df_by_user = _get_combined_stay(df_by_user)
+
+    return df_by_user
+
+
+def _run(args):
+    df_by_user, spatial_constraint, dur_constraint = args
+
+    if dur_constraint == -1:
+        df_by_user = _run_for_user(df_by_user, spatial_constraint)
+
+    else:
+        df_by_user = _run_for_user(df_by_user, spatial_constraint, dur_constraint)
+
+    return df_by_user
+
+
+if __name__ == "__main__":
```

Comment:
what's the use of this block? Is this just missing the call to `incremental_clustering`?

Reply:
Removed, not required